File: src/main/java/sawyern/cookiebot/commands/GiveCookieCommand.java
package sawyern.cookiebot.commands;

import com.google.common.collect.Sets;
import discord4j.core.event.domain.message.MessageCreateEvent;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import sawyern.cookiebot.models.dto.GiveCookieDto;
import sawyern.cookiebot.exception.CookieException;
import sawyern.cookiebot.services.CookieService;

import java.text.MessageFormat;
import java.util.List;
import java.util.Set;

@Component
@RequiredArgsConstructor
public class GiveCookieCommand extends MessageCreateEventBotCommand {

    private final CookieService cookieService;

    @Override
    public String getCommand() {
        return "givecookie";
    }

    @Override
    public Set<Integer> getAllowedNumArgs() {
        return Sets.newHashSet(1, 2);
    }

    @Override
    public void execute(MessageCreateEvent event, List<String> args) throws CookieException {
        GiveCookieDto giveCookieDto = new GiveCookieDto();
        giveCookieDto.setSenderId(botUtil.getMember(event).getId().asString());

        String senderUser = botUtil.getMember(event).getUsername();
        String recieverUser = null;

        if (args.size() == 1) {
            recieverUser = args.get(0);
            if (recieverUser.equalsIgnoreCase(botUtil.getMember(event).getUsername()))
                throw new CookieException("Can't give cookies to yourself.");

            // first argument is username
            giveCookieDto.setRecieverId(botUtil.getIdFromUser(event, args.get(0)));
            // give one cookie if not specified
            giveCookieDto.setNumCookies(1);
        }
        else if (args.size() == 2) {
            recieverUser = args.get(1);
            if (recieverUser.equalsIgnoreCase(botUtil.getMember(event).getUsername()))
                throw new CookieException("Can't give cookies to yourself.");

            // second argument is username
            giveCookieDto.setRecieverId(botUtil.getIdFromUser(event, args.get(1)));
            // first argument is numCookies
            giveCookieDto.setNumCookies(botUtil.parsePositiveIntArgument(args.get(0)));
        }

        cookieService.giveCookieTo(giveCookieDto);

        int senderTotal = cookieService.getAllCookiesForAccount(giveCookieDto.getSenderId());
        int recieverTotal = cookieService.getAllCookiesForAccount(giveCookieDto.getRecieverId());

        botUtil.sendMessage(event, MessageFormat.format(
                "Successfully transferred cookies. {0}: {1}, {2}: {3}",
                senderUser,
                String.valueOf(senderTotal),
                recieverUser,
                String.valueOf(recieverTotal))
        );
    }
}

File: src/main/kotlin/sawyern/cookiebot/commands/GiveCookieCommand.kt
package sawyern.cookiebot.commands

import discord4j.core.event.domain.message.MessageCreateEvent
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import sawyern.cookiebot.exception.CookieException
import sawyern.cookiebot.services.BotUtilService
import sawyern.cookiebot.services.CookieService
import java.text.MessageFormat

@Component
class GiveCookieCommand @Autowired constructor(
        botUtilService: BotUtilService,
        private val cookieService: CookieService
): MessageCreateEventBotCommand(botUtilService) {
    override fun getAllowedNumArgs(): Set<Int> {
        return setOf(1, 2)
    }

    override fun getCommand(): String {
        return "givecookie"
    }

    override fun getHelpText(): String {
        return "give cookie to {receiver}"
    }

    override fun execute(event: MessageCreateEvent, args: List<String>) {
        val senderUser = botUtilService.getMember(event)
        val receiverUser: String = args[0]
        val receiverId: String = botUtilService.getIdFromUser(event, args[0])
        var numCookies = 1

        if (receiverUser.equals(botUtilService.getMember(event).username, ignoreCase = true))
            throw CookieException("Can't give cookies to yourself.")

        if (args.size == 2)
            numCookies = botUtilService.parseIntArgument(args[1], true)

        cookieService.giveCookieTo(senderUser.id.asString(), receiverId, numCookies)

        val senderTotal = cookieService.getCookiesForAccount(senderUser.id.asString())
        val receiverTotal = cookieService.getCookiesForAccount(receiverId)

        botUtilService.sendMessage(event, MessageFormat.format(
                "Successfully transferred cookies. {0}: {1}, {2}: {3}",
                senderUser.username,
                senderTotal.toString(),
                receiverUser,
                receiverTotal.toString())
        )
    }
}